src/null-safety/unsound-null-safety.md

```diff
@@ -0,0 +1,179 @@
+
+[migration tool]: /null-safety/migration-guide#step2-migrate
+
+
+We recommend that you **first migrate leaf libraries** —
+libraries that don't import other files from the package.
+Then migrate libraries that directly depend on the leaf libraries.
+End by migrating the libraries that have the most
+intra-package dependencies.
+
+For example, say you have a `lib/src/util.dart` file
+that imports other (null-safe) packages and core libraries,
+but that doesn't have any `import '<local_path>'` directives.
+Consider migrating `util.dart` first,
+and then migrating files that depend only on `util.dart`.
+If any libraries have cyclic imports
+(for example, A imports B which imports C, and C imports A),
+consider migrating those libraries together.
+
+To migrate a package by hand, follow these steps:
+
+1. Edit the package's `pubspec.yaml` file,
+   setting the minimum SDK constraint to `2.12.0-0`:
+   ```yaml
+environment:
+  sdk: '>=2.12.0-0 <3.0.0'
+```
+
+2. Regenerate the [package configuration file][]:
+
+   ```terminal
+   $ dart pub get
+   ```
+
+   [package configuration file]: https://github.com/dart-lang/language/blob/master/accepted/future-releases/language-versioning/package-config-file-v2.md
+
+   Running `dart pub get` with a lower SDK constraint of `2.12.0-0`
+   sets the default language version of
+   every library in the package to 2.12,
+   opting them all in to null safety.
+
+3. Open the package in your IDE. <br>
+   You're likely to see a lot of analysis errors.
+   That's OK.
+
+4. Add a [language version comment][] to the top of
+   any Dart files that you don't want to consider during your current migration:
+   ```dart
+// @dart=2.9
+```
+
+   Using language version 2.9 for a library that's in a 2.12 package
+   can reduce analysis errors (red squiggles) coming from unmigrated code.
+   However, **unsound null safety reduces the
+   information the analyzer can use.**
+   For example, the analyzer might assume a
+   parameter type is non-nullable,
+   even though a 2.9 file might pass in a null value.
+
+5. Migrate the code of each Dart file,
+   using the analyzer to identify static errors. <br>
+   Eliminate static errors by adding `?`, `!`, `required`, and `late`,
+   as needed.
+
+
+## Testing or running mixed-version programs
+
+To test or run mixed-version code,
+you need to disable sound null safety.
+You can do this in two ways:
+
+* Disable sound null safety using the `--no-sound-null-safety` flag.
+  Example:
+
+  ```terminal
+  $ dart --no-sound-null-safety run
+  ```
+
+* Alternatively, set the language version in the entrypoint —
+  the file that contains `main()` function — to 2.9.
+  In Flutter apps, this file is often named `lib/main.dart`.
+  In command-line apps, this file is often named `bin/<packageName>.dart`.
+  You can also opt out files under `test`,
+  because they are also entrypoints.
+  Example:
+
+  ```dart
+  // @dart=2.9
+  import 'src/my_app.dart';
+
+  main() {
+    //...
+  }
+  ```
+
```

Comment:
I think it's worth re-emphasizing here that this is not a desirable end state.  I'd say something like:

```
While opting out your test files may be useful to allow you to run tests **during** your incremental migration process, doing so means that you won't be testing your code with full null safety enabled.  Therefore, it's important to opt your tests back in once you have finished the incremental migration of your libraries.
```

Reply:
I'll do this in a follow-up PR

Reply:
https://github.com/dart-lang/site-www/pull/2770